### PR TITLE
HPCC-19493 Add access to activity context for other embedded languages

### PR DESCRIPTION
**Pull request by richardkchapman**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
Added test code, similar to the existing cpp embed test case.

### CODE REVIEW

**Comment by richardkchapman** [comment]
@ghalliday For initial review - only covers Python 2/3 at present, but if the approach looks right I will think about others

File: testing/regress/ecl/mysqlembed.ecl [hunk 1]
```diff
@@ -80,6 +80,10 @@ dataset(childrec) testMySQLDS2() := EMBED(mysql : server(myServer),user(myUser),
   SELECT [] from tbl1 where u1='Straße';
 ENDEMBED;
 
+streamed dataset(childrec) testMySQLDS2a() := EMBED(mysql : server(myServer),user(myUser),database(myDB),PROJECTED('[]'), activity, local(true))
+  SELECT [] from tbl1 where (NOT __LOCAL__) OR (__NUMSLAVES__<2) OR (value = __SLAVE__+1);
```
Comment (ghalliday) [inline testing/regress/ecl/mysqlembed.ecl]: Will the last element  work on a 4 way thor?

**Comment by ghalliday** [comment]
@richardkchapman one question.  Generally looks ok.

File: testing/regress/ecl/mysqlembed.ecl (hunk 1 above)
Comment (richardkchapman) [reply testing/regress/ecl/mysqlembed.ecl]: I think so, as there are only 2 items in the dataset.

**Comment by richardkchapman** [comment]
@ghalliday Note I added another commit for javascript. I should probably try to do R, but not sure I have the R skills (or the energy)

File: testing/regress/ecl/embedjs.ecl [hunk 2]
```diff
@@ -101,6 +101,32 @@ val[2] = t;
...
+    var last = first + rowsPerPart;
+  
+    var names = [ "Gavin", "Richard", "John", "Bart" ];
+    var ds = [ ];
+    while (first < last)
+    {
+        ds.push( { id : first, name: names[first % 4] });
+        first += 1;
+    }
+    ds;
```
Comment (GordonSmith) [inline testing/regress/ecl/embedjs.ecl]: How does "ds;" magically get returned?

Comment (richardkchapman) [reply testing/regress/ecl/embedjs.ecl]: I don't remember how the embedded JS works - I just copied the other examples in this file...

**Review by ghalliday — COMMENTED** [review]
A few comments.  Looks good though.

File: system/jlib/jstring.cpp [hunk 3]
```diff
@@ -947,6 +947,33 @@ StringBuffer &replaceString(StringBuffer & result, size32_t lenSource, const cha
     return result;
 }
 
+StringBuffer &replaceStringNoCase(StringBuffer & result, size32_t lenSource, const char *source, size32_t lenOldStr, const char* oldStr, size32_t lenNewStr, const char* newStr)
+{
+    if (lenSource)
+    {
+        size32_t left = lenSource;
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply system/jlib/jstring.cpp]: This is just a copy of the case-sensitive code...

**Comment by richardkchapman** [comment]
@ghalliday Please look again

**Pull request merged**